### PR TITLE
[LS] Implement a simulator for language server

### LINKED ISSUE #32658: Implement a simulator to simulate extended coding sessions with language server
**Description:**
With #32020, we have a requirement to automate finding issues that occur when a user is using vscode for an extended period. Mainly we are focused in finding memory leaks and concurrency issues. Need to implement a simulator to simulate the behavior of a user using vscode to code a ballerina program with the help of language server,

**Describe your task(s)**
See description

**Related Issues (optional):**
#32020

### PR DESCRIPTION
**Pull request by IMS94**
## Purpose
This PR introduces a set of classes to simulate the behavior of LS and vscode combination. It uses the language server protocol's messages to send document changes to the LS to simulate a user writing some code character by character.

This is will be helpful in finding issues like #32020.
Fixes #32658

## Approach
Introduced classes `Editor`, `EditorTab`, `code snippet generators, etc to simulate the vscode behavior.

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: tests/language-server-simulator/src/main/java/org/ballerinalang/langserver/simulator/Editor.java [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+        });
+    }
+
+    public void closeTab(EditorTab tab) {
+        tabs.remove(tab);
+    }
+
+    public void close() {
+        this.languageServer.shutdown();
+        tabs.forEach(tab -> tab.close());
```
Comment (malinthar) [inline tests/language-server-simulator/src/main/java/org/ballerinalang/langserver/simulator/Editor.java:93]:
```suggestion
        tabs.forEach(EditorTab::close);
```

File: tests/language-server-simulator/src/main/java/org/ballerinalang/langserver/simulator/Editor.java [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+            if (filePath.equals(tab.filePath())) {
+                tab.close();
+                return true;
+            }
+            return false;
+        });
+    }
+
+    public void closeTab(EditorTab tab) {
+        tabs.remove(tab);
```
Comment (malinthar) [inline tests/language-server-simulator/src/main/java/org/ballerinalang/langserver/simulator/Editor.java:88]: Shouldn't we consider the `activetab` reference in this case. If the provided tab is the active tab, shouldn't we set it to something else.

File: tests/language-server-simulator/src/main/java/org/ballerinalang/langserver/simulator/EditorTab.java [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+            TestUtil.didChangeDocument(this.endpoint, this.filePath, textDocument.toString());
+        }
+    }
+
+    /**
+     * Check if the document in this instance is similar to that is in the language server.
+     *
+     * @return
+     */
+    private boolean isDocumentInSync() {
```
Comment (malinthar) [inline tests/language-server-simulator/src/main/java/org/ballerinalang/langserver/simulator/EditorTab.java:148]: This method is always inverted. Shall we send the negation?

**Review by malinthar — CHANGES_REQUESTED** [review]
Changes look good!. Have few minor change requests/

File: tests/language-server-simulator/build.gradle [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+            }
+        }
+    }
+}
+
+task runLSSimulator(type: JavaExec) {
+    dependsOn unpackBalTestProject
+    dependsOn loadDistributionCache
+    systemProperty "ballerina.home", "$buildDir/"
+    systemProperty "experimental", "true"
```
Comment (mohanvive) [inline tests/language-server-simulator/build.gradle:115]: Remove `experimental` flag.

**Review by mohanvive — APPROVED** [review]
LGTM. Let's fix the minor issues and get this merged.

File: tests/language-server-simulator/build.gradle [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+
+test {
+    dependsOn copyCompilerPluginJars
+    dependsOn loadDistributionCache
+    systemProperty "ballerina.home", "$buildDir/"
+    systemProperty "experimental", "true"
+    systemProperty "ballerina.version", project.version
+    useTestNG() {
+        suites 'src/test/resources/testng.xml'
+    }
```
[the remark this answers is not in the document]
Comment (IMS94) [reply tests/language-server-simulator/build.gradle]: Yes. Removed this section.

File: tests/language-server-simulator/src/main/java/org/ballerinalang/langserver/simulator/EditorTab.java (hunk 3 above)
Comment (IMS94) [reply tests/language-server-simulator/src/main/java/org/ballerinalang/langserver/simulator/EditorTab.java]: Done

File: tests/language-server-simulator/build.gradle [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+plugins {
+    id 'de.undercouch.download' version '4.1.1'
```
[the remark this answers is not in the document]
Comment (IMS94) [reply tests/language-server-simulator/build.gradle]: We can use 4.1.2. Done!

File: tests/language-server-simulator/build.gradle (hunk 4 above)
Comment (IMS94) [reply tests/language-server-simulator/build.gradle]: Done!

File: tests/language-server-simulator/src/main/java/org/ballerinalang/langserver/simulator/Editor.java (hunk 2 above)
Comment (IMS94) [reply tests/language-server-simulator/src/main/java/org/ballerinalang/langserver/simulator/Editor.java:92]: Agreed. Addressed.

**Review by mohanvive — APPROVED** [review]

**Review by malinthar — APPROVED** [review]
LGTM!

**Review by mohanvive — APPROVED** [review]

**Pull request merged**